Funcky/Extensions/EnumerableExtensions/AdjacentGroupBy.cs

```diff
@@ -0,0 +1,242 @@
+        /// <typeparam name="TResult">The type of the result value returned by resultSelector.</typeparam>
+        /// <param name="source">The source sequence.</param>
+        /// <param name="keySelector">A function to extract the key for each element.</param>
+        /// <param name="elementSelector">A function to map each source element to an element in the IGrouping{TKey,TElement}.</param>
+        /// <param name="resultSelector">A function to create a result value from each group.</param>
+        /// <param name="comparer">An IEqualityComparer{T} to compare keys.</param>
+        /// <returns>A collection of elements of type TResult where each element represents a projection over a group and its key.</returns>
+        [Pure]
+        public static IEnumerable<TResult> AdjacentGroupBy<TSource, TKey, TElement, TResult>(
+            this IEnumerable<TSource> source,
+            Func<TSource, TKey> keySelector,
+            Func<TSource, TElement> elementSelector,
+            Func<TKey, IList<TElement>, TResult> resultSelector,
+            IEqualityComparer<TKey> comparer)
+        {
+            using var enumerator = source.GetEnumerator();
+
+            if (!enumerator.MoveNext())
+            {
+                yield break;
+            }
+
+            var group = new List<TElement>() { elementSelector(enumerator.Current) };
+            var key = keySelector(enumerator.Current);
+
+            while (enumerator.MoveNext())
+            {
+                if (comparer.Equals(key, keySelector(enumerator.Current)))
+                {
+                    group.Add(elementSelector(enumerator.Current));
+                }
+                else
+                {
+                    yield return resultSelector(key, group);
+                    group = new List<TElement>() { elementSelector(enumerator.Current) };
+                    key = keySelector(enumerator.Current);
+                }
+            }
+
+            yield return resultSelector(key, group);
+        }
+
+        internal static Grouping<TKey, TElement> CreateGrouping<TKey, TElement>(TKey key, IList<TElement> elements)
+        {
+            return new Grouping<TKey, TElement>(key, elements);
+        }
+
+        public class Grouping<TKey, TElement> : IGrouping<TKey, TElement>, IList<TElement>
+        {
+            private readonly IList<TElement> _elements;
+
+            internal Grouping(TKey key, IList<TElement> elements)
+            {
+                Key = key;
+                _elements = elements;
+            }
+
+            public TKey Key { get; }
+
+            public int Count
+                => _elements.Count;
+
+            public bool IsReadOnly
+                => true;
+
+            public TElement this[int index]
+            {
+                get => _elements[index];
+                set => throw new NotImplementedException();
+            }
+
+            public void Add(TElement element)
+                => _elements.Add(element);
```

Comment:
Shouldn't this also throw a `NotSupportedException`, since the grouping is read only?

Reply:
Addressed in: 0c093e77
Split interface ICollection<TElement>.Add which throws NotSupportedException from internal Add function I need as an interface.